independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/bean/types/GenericBeanTypesTest.java

```diff
@@ -0,0 +1,52 @@
+package io.quarkus.arc.test.bean.types;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.InjectableBean;
+import io.quarkus.arc.test.ArcTestContainer;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.util.Iterator;
+import java.util.Set;
+import javax.enterprise.context.Dependent;
+import javax.inject.Named;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class GenericBeanTypesTest {
+    @RegisterExtension
+    ArcTestContainer container = new ArcTestContainer(MyBean.class);
+
+    @Test
+    public void customGenericBean() {
+        InjectableBean<Object> bean = Arc.container().instance("myBean").getBean();
+        Set<Type> types = bean.getTypes();
+
+        assertEquals(3, types.size());
+        assertTrue(types.contains(Object.class));
+        for (Type type : types) {
+            if (type instanceof ParameterizedType) {
+                Type genericClass = ((ParameterizedType) type).getRawType();
+                assertTrue(MyBean.class.equals(genericClass) || Iterable.class.equals(genericClass));
+
+                Type[] typeArgs = ((ParameterizedType) type).getActualTypeArguments();
+                assertEquals(1, typeArgs.length);
+                Type typeArg = typeArgs[0];
+                assertTrue(typeArg instanceof TypeVariable);
+                assertEquals("T", ((TypeVariable) typeArg).getName());
```

Comment:
Shouldn't we test the bounds here as well?

Reply:
Good point.

Reply:
Done.